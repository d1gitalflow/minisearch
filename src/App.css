.button:hover > svg > g {
  fill: #ffffff;
  stroke: #a79d9d;
  }