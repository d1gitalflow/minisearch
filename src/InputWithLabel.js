import { useEffect, useRef } from "react";
//abstracted component / reusable component
export const InputWithLabel = ({ id, label, value, onInputChange, type = 'text', children, isFocus }) => {

    const handleChange = (ev) => {
      onInputChange(ev)
    }
  
    const inputRef = useRef();
    useEffect(() => {
      if (isFocus && inputRef.current) {
        inputRef.current.focus();
      }
    }, [isFocus])
    return (
      <>
        <label htmlFor={id} >{children}</label>
        
        <input
          id={id}
          type={type}
          value={value}
          onChange={handleChange}
  
          ref={inputRef}
        ></input>
      </>
    )
  }