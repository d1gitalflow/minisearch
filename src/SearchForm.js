import {InputWithLabel} from './InputWithLabel.js';
export const SearchForm = ({ onSearchSubmit, onSearchInput, searchTerm }) => {
    return (
      <form onSubmit={onSearchSubmit}>
        <InputWithLabel
          id={'search'}
          label={'Search'}
          value={searchTerm}
          onInputChange={onSearchInput}
          isFocus={true}
        >
          <strong>Search:</strong>
        </InputWithLabel>
        <button
          
          type="submit"
          disabled={!searchTerm}
        >SubmitYo
        </button>
      </form>
    )
  }