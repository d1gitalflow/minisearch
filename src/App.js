import { useState, useEffect, useRef, useReducer, useCallback, useMemo } from "react";
import axios from 'axios/index.js';
import { SearchForm } from "./SearchForm";
import { InputWithLabel } from "./InputWithLabel";
import { List ,Item} from "./List";

//utilities such as sorting arrays
import { sortBy } from 'lodash';


//custom hook (we merge the two react core hooks useState and useEffect)
const useSemiPersistentState = (key, initialState) => {

  const isMounted = useRef(false); //current property initialized with 'false'

  const [value, setValue] = useState(
    localStorage.getItem(key) || initialState
  );
  useEffect(() => {
    if (!isMounted.current) {
      isMounted.current = true;  //isMounted.current if set to true, it survives full page re-render and if
      // exists it will set newItem the next updates avoiding the first re-render
    } else {
      localStorage.setItem(key, value);
    }

  }, [value, key]);

  return [value, setValue];
};

//action parameter =>{type:'type',payload:value}
//state parameter: init from useReducer(), updated from useReducer()
const storiesReducer = (state, action) => {
  switch (action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

export const SORTS = {
  NONE: (list) => list,
  TITLE: (list) => sortBy(list, 'title'),
  AUTHOR: (list) => sortBy(list, 'author'),
  COMMENT: (list) => sortBy(list, 'num_comments').reverse(),
  POINT: (list) => sortBy(list, 'points').reverse(),
};

const REMOVE_STORY = 'REMOVE_STORY';
const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';

const getSumComments = (stories) => {
  return stories.data.reduce(
    (result, value) => result + value.num_comments,
    0
  );
};
const extractSearchTerm = (url) => url.replace(API_ENDPOINT, '');
const getLastSearches = (urls) =>urls.slice(-6)//[a,b,c,d,e,f]
                                .slice(0,-1)   //[a,b,c,d,e] f - is the most recent it doesn't display
                                .map((url) => extractSearchTerm(url)); //-5 is the last five

const getUrl = (searchTerm) => `${API_ENDPOINT}${searchTerm}`;

const App = () => {


  //const [searchTerm, setSearchTerm] = useState(localStorage.getItem('search') || 'React');
  //custom hook
  const [searchTerm, setSearchTerm] = useSemiPersistentState('search', 'React');

  const [urls, setUrls] = useState([getUrl(searchTerm)]);

  const handleSearch = (searchTerm) => {
    const url = getUrl(searchTerm);
    setUrls(urls.concat(url));
    };

  //update the setUrl
  const handleLastSearch = (searchTerm) => {
    handleSearch(searchTerm);
  }

  const lastSearches = getLastSearches(urls);

 


  //useReducer hook, we pass the reducer function + and the initial state
  //in useReducer we call the state updater function 'dispatcherBlaBla'
  const [stories, dispatchStories] = useReducer(storiesReducer, { data: [], isLoading: false, isError: false });


  const handleSearchInput = (event) => {
    setSearchTerm(event.target.value)
  }

  //submit the searchTerm state
  const handleSearchSubmit = (event) => {
    handleSearch(searchTerm);

    //<button type="submit"> to prevent from reloading the page
    //"prevent default behavior"
    event.preventDefault();
  }

  const runComputation = useMemo(() => getSumComments(stories), [stories])


  const handleFetchStories = useCallback(async () => {
    if (!searchTerm) return;
    dispatchStories({ type: 'STORIES_FETCH_INIT' });

    try {
      //resolves in the latest url
      const lastUrl = urls[urls.length - 1];
      //fetches last url from urls present state
      const result = await axios.get(lastUrl);
      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS',
        payload: result.data.hits,
      });
    }
    catch {
      dispatchStories({
        type: 'STORIES_FETCH_FAILURE'
      })
    }

    //eslint-disable-next-line
  }, [urls])

  useEffect(() => {
    handleFetchStories()
  }, [handleFetchStories]);



  //set if passed story doesn't equal story.objectID
  //call happens on Item component function
  const handleRemoveStory = useCallback((item) => {

    //we pass the item as a payload
    dispatchStories({
      type: REMOVE_STORY,
      payload: item
    })

  }, [])




  return (
    <div >
      <h1>My bleeing edge react testing ground with {runComputation} comments!</h1>

      <SearchForm
        onSearchSubmit={handleSearchSubmit}
        onSearchInput={handleSearchInput}
        searchTerm={searchTerm}

      />

      {lastSearches.map((searchTerm, index) => {
        return (
          <button
            key={searchTerm + index}
            type="button"
            onClick={() => handleLastSearch(searchTerm)}
          >
            {searchTerm}
          </button>
        )
      })}


      {stories.isError && <p>Oops something went wrong...</p>}
      {stories.isLoading ? (<p>Loading...</p>) : (
        <List
          list={stories.data}
          onRemoveItem={handleRemoveStory}
        />
      )}



    </div>

  )
}













export default App;

export { storiesReducer, SearchForm, InputWithLabel, List, Item };