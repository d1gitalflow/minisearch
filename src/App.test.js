import * as React from 'react'
import { render, screen, fireEvent, act } from '@testing-library/react';
import App, { storiesReducer } from './App';
import { InputWithLabel } from './InputWithLabel';
import { List, Item } from './List';
import { SearchForm } from './SearchForm';
import axios from 'axios';
jest.mock('axios');







describe('something truthy and falsy', () => {
  test('true to be true', () => {
    expect(true).toBe(true);
  })
  test('false to be false', () => {
    expect(false).toBe(false);
  })

})


const storyOne = {
  title: 'React',
  url: 'https://reactjs.org/',
  author: 'Jordan Walke',
  num_comments: 3,
  points: 4,
  objectID: 0,
}

const storyTwo = {
  title: 'Redux',
  url: 'https://redux.js.org/',
  author: 'Dan Abramov, Andrew Clark',
  num_comments: 2,
  points: 5,
  objectID: 1,
};

const stories = [storyOne, storyTwo];

describe('App', () => {
  test('succeeds fetching data', async () => {
    //to be resolved promise
    const promise = Promise.resolve({
      data: {
        hits: stories
      }
    })
    axios.get.mockImplementationOnce(() => promise);

    render(<App />);
    expect(screen.getByText(/Loading/)).toBeInTheDocument();

    //promise resolves inside act
    await act(() => promise);
    expect(screen.queryByText(/Loading/)).toBeNull();

    expect(screen.getByText('React')).toBeInTheDocument();
    expect(screen.getByText('Redux')).toBeInTheDocument();
    expect(screen.getAllByText('Submit').length).toBe(2);

  })

  test('fails fetching data', async () => {
    const promise = Promise.reject();
    axios.get.mockImplementationOnce(() => promise);

    render(<App />);

    expect(screen.getByText(/Loading/)).toBeInTheDocument();

    try {
      await act(() => promise);
    } catch {
      expect(screen.queryByText(/Loading/)).toBeNull();
      expect(screen.queryByText(/went wrong/)).toBeInTheDocument();
    }
  })

  test('removes a story', async () => {
    const reactPromise = Promise.resolve({
      data: {
        hits: stories,
      },
    });
    const anotherStory = {
      title: 'JavaScript',
      url: 'https://en.wikipedia.org/wiki/JavaScript',
      author: 'Brendan Eich',
      num_comments: 15,
      points: 10,
      objectID: 3,
    };
    const javascriptPromise = Promise.resolve({
      data: {
        hits: [anotherStory],
      },
    });
    axios.get.mockImplementation((url) => {
      if (url.includes('React')) {
        return reactPromise;
      }
      if (url.includes('JavaScript')) {
        return javascriptPromise;
      }
      throw Error();
    });

    //first render
    render(<App />);
    //first data fetch
    await act(() => reactPromise)

    expect(screen.queryByDisplayValue('React')).toBeInTheDocument();
    expect(screen.queryByDisplayValue('JavaScript')).toBeNull();
    expect(screen.queryByText('Jordan Walke')).toBeInTheDocument();
    expect(screen.queryByText('Brendan Eich')).toBeNull();

    fireEvent.change(screen.queryByDisplayValue('React'), { target: { value: 'JavaScript' } })
    expect(
      screen.queryByDisplayValue('JavaScript')
    ).toBeInTheDocument();

    fireEvent.submit(screen.getByRole('button', { name: 'SubmitYo' }));
    await act(() => javascriptPromise);

    expect(screen.queryByText('Jordan Walke')).toBeNull();
    expect(
      screen.queryByText('Dan Abramov, Andrew Clark')
    ).toBeNull();
    expect(screen.queryByText('Brendan Eich')).toBeInTheDocument();
  });

})



describe('storiesReducer()', () => {
  //extrapolate/extend the test cases to one per action from storiesReducer
  test('REMOVE_STORY', () => {
    const action = { type: 'REMOVE_STORY', payload: storyOne };
    const state = { data: stories, isLoading: false, isError: false }
    const newState = storiesReducer(state, action);
    const expectedState = { data: [storyTwo], isLoading: false, isError: false }
    expect(newState).toStrictEqual(expectedState)
  })

  test('STORIES_FETCH_INIT', () => {
    const action = { type: 'STORIES_FETCH_INIT' };
    const state = { data: [], isLoading: false, isError: false }
    const newState = storiesReducer(state, action);
    const expectedState = { data: [], isLoading: true, isError: false };
    expect(newState).toStrictEqual(expectedState);
  })

  test('STORIES_FETCH_SUCCESS', () => {
    const action = { type: 'STORIES_FETCH_SUCCESS', payload: stories };
    const state = { data: stories, isLoading: false, isError: false }
    const newState = storiesReducer(state, action);
    const expectedState = { data: [storyOne, storyTwo], isLoading: false, isError: false }
    expect(newState).toStrictEqual(expectedState);
  })

  test('STORIES_FETCH_FAILURE', () => {
    const action = { type: 'STORIES_FETCH_FAILURE' };
    const state = { data: stories, isLoading: false, isError: false }
    const newState = storiesReducer(state, action);
    const expectedState = { data: stories, isLoading: false, isError: true }
    expect(newState).toStrictEqual(expectedState);
  })
})


describe('Item', () => {
  test('renders all properties', () => {
    render(<Item item={storyOne} />);

    expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
    expect(screen.getByText('React')).toHaveAttribute('href', 'https://reactjs.org/');

  })

  test('renders a clickable dissmiss button', () => {
    render(<Item item={storyOne} />);
    expect(screen.getByRole('button')).toBeInTheDocument();
  })

  test('clicking the dismiss button calls the callback handler', () => {
    const handleRemoveItem = jest.fn().mockImplementationOnce((item) => item);
    render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);

    fireEvent.click(screen.getByRole('button'));

    expect(handleRemoveItem.mock.results[0].value).toStrictEqual({
      title: 'React',
      url: 'https://reactjs.org/',
      author: 'Jordan Walke',
      num_comments: 3,
      points: 4,
      objectID: 0
    });
  })

  test('renders snapshot',()=>{
    const handleRemoveItem = jest.fn();
    const {container} = render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />)
    expect(container.firstChild).toMatchSnapshot();
  })
})

describe('SearchForm', () => {
  const searchFormProps = {
    searchTerm: 'React',
    onSearchSubmit: jest.fn(),
    onSearchInput: jest.fn(),
  }

  test('renders the input field with its value', () => {
    render(<SearchForm {...searchFormProps} />)
    expect(screen.getByDisplayValue('React')).toBeInTheDocument();
  })

  test('renders the correct label', () => {
    render(<SearchForm {...searchFormProps} />);
    expect(screen.getByLabelText('Search:')).toBeInTheDocument();
  });

  test('calls onSearchInput on input field change', () => {
    render(<SearchForm {...searchFormProps} />);
    fireEvent.change(screen.getByDisplayValue('React'), { target: { value: 'Redux' } })
    expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1)

  })

  test('calls onSearchSubmit on button submit click', () => {
    render(<SearchForm {...searchFormProps} />);
    fireEvent.submit(screen.getByRole('button'), { name: 'SubmitYo' });
    expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
  })

  test('renders snapshot',()=>{
    const {container} = render(<SearchForm {...searchFormProps}/>);
    expect(container.firstChild).toMatchSnapshot();
  })
})






