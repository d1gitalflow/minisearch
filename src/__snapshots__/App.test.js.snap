// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders snapshot 1`] = `
<li>
  <span
    style="width: 40%;"
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    style="width: 30%;"
  >
    Jordan Walke
  </span>
  <span
    style="width: 10%;"
  >
    3
  </span>
  <span
    style="width: 10%;"
  >
    4
  </span>
  <span
    style="width: 10%;"
  >
    <button
      type="button"
    >
      Submit
    </button>
  </span>
</li>
`;

exports[`SearchForm renders snapshot 1`] = `
<form>
  <label
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  <input
    id="search"
    type="text"
    value="React"
  />
  <button
    type="submit"
  >
    SubmitYo
  </button>
</form>
`;
